### PR TITLE
Upgrade zlib to 1.3.1

### PR DESCRIPTION
**Pull request by carlossanlop**

### CODE REVIEW

File: src/native/external/zlib-version.txt [hunk 1]
```diff
@@ -1,17 +1,9 @@
...
 
-We have also cherry-picked into our local copy:
-
-- https://github.com/madler/zlib/commit/e554695638228b846d49657f31eeff0ca4680e8a
-
-  This patch only affects memLevel 9 compression. .NET doesn't currently use this
-  memLevel, but we'll take this patch out of an abundance of caution just in case
-  we enable this functionality in a future release.
-
-We have also applied the custom patches under the patches/zlib folder.
```
Comment (GrabYourPitchforks) [inline src/native/external/zlib-version.txt:17]: Was removing this line intentional?

**Review by GrabYourPitchforks — APPROVED** [review]
LGTM, modulo what appears to be a stray readme deletion. I diffed the files against the zlib repo and the only changes are what are documented in our readme file.

File: src/native/external/zlib-version.txt (hunk 1 above)
Comment (carlossanlop) [reply src/native/external/zlib-version.txt:17]: It was removed intentionally until I learned about the patch and applied it, but forgot to restore it. I should bring it back in all the PRs.

File: src/native/external/zlib-version.txt [hunk 2]
```diff
@@ -1,17 +1,9 @@
...
+The file zlib2ansi was removed in this version release.
```
Comment (jkotas) [inline src/native/external/zlib-version.txt:9]:
```suggestion
```
This was mentioned in zlib release notes. I do not think we need to mention it here again - it does not how zlib is used by .NET.

Comment (carlossanlop) [inline src/native/external/zlib-version.txt:9]:
```suggestion
We have also applied the custom patches under the patches/zlib folder.
```

**Review by GrabYourPitchforks — APPROVED** [review]

**Comment by carlossanlop** [comment]
@ericstj @GrabYourPitchforks @artl93 

There was no significant change in the performance comparison of the previous commit (e9e33e1) vs the commit that updated zlib (e5179e7):


| Method     | Commit        | level   | file             | Mean       | Error    | StdDev   | Median     | Min        | Max        | Ratio | MannWhitney(5%) | RatioSD | Allocated | Alloc Ratio |
|----------- |----------- |-------- |----------------- |-----------:|---------:|---------:|-----------:|-----------:|-----------:|------:|---------------- |--------:|----------:|------------:|
| Compress   | e9e33e1 (Old) | Optimal | alice29.txt      | 5,147.3 us | 13.78 us | 12.21 us | 5,144.9 us | 5,131.6 us | 5,172.8 us |  1.00 | Base            |    0.00 |   8.26 KB |        1.00 |
| Compress   | e5179e7 (New) | Optimal | alice29.txt      | 5,147.0 us |  8.87 us |  7.86 us | 5,149.1 us | 5,134.5 us | 5,161.1 us |  1.00 | Same            |    0.00 |   8.26 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Decompress | e9e33e1 (Old) | Optimal | alice29.txt      |   506.8 us |  2.28 us |  2.02 us |   506.0 us |   503.9 us |   510.6 us |  1.00 | Base            |    0.01 |   8.27 KB |        1.00 |
| Decompress | e5179e7 (New) | Optimal | alice29.txt      |   506.7 us |  2.18 us |  1.93 us |   506.1 us |   504.0 us |   510.0 us |  1.00 | Same            |    0.01 |   8.27 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Compress   | e9e33e1 (Old) | Optimal | sum              |   739.6 us |  1.57 us |  1.39 us |   738.9 us |   738.0 us |   742.3 us |  1.00 | Base            |    0.00 |   8.26 KB |        1.00 |
| Compress   | e5179e7 (New) | Optimal | sum              |   738.6 us |  1.77 us |  1.48 us |   738.0 us |   737.3 us |   742.3 us |  1.00 | Same            |    0.00 |   8.26 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Decompress | e9e33e1 (Old) | Optimal | sum              |   144.4 us |  2.43 us |  2.15 us |   145.2 us |   139.7 us |   146.6 us |  1.00 | Base            |    0.02 |   8.27 KB |        1.00 |
| Decompress | e5179e7 (New) | Optimal | sum              |   144.2 us |  2.15 us |  1.79 us |   144.8 us |   140.2 us |   146.1 us |  1.00 | Same            |    0.02 |   8.27 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Compress   | e9e33e1 (Old) | Optimal | TestDocument.pdf | 3,167.1 us |  8.11 us |  6.77 us | 3,166.5 us | 3,158.3 us | 3,182.5 us |  1.00 | Base            |    0.00 |   8.26 KB |        1.00 |
| Compress   | e5179e7 (New) | Optimal | TestDocument.pdf | 3,165.2 us | 15.75 us | 14.73 us | 3,165.6 us | 3,146.6 us | 3,192.7 us |  1.00 | Same            |    0.00 |   8.26 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Decompress | e9e33e1 (Old) | Optimal | TestDocument.pdf |   397.3 us |  1.08 us |  0.90 us |   397.2 us |   395.7 us |   398.8 us |  1.00 | Base            |    0.00 |   8.27 KB |        1.00 |
| Decompress | e5179e7 (New) | Optimal | TestDocument.pdf |   398.1 us |  1.15 us |  1.02 us |   397.6 us |   396.8 us |   400.3 us |  1.00 | Same            |    0.00 |   8.27 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Compress   | e9e33e1 (Old) | Fastest | alice29.txt      | 1,722.5 us |  3.19 us |  2.83 us | 1,721.9 us | 1,719.2 us | 1,729.7 us |  1.00 | Base            |    0.00 |   8.26 KB |        1.00 |
| Compress   | e5179e7 (New) | Fastest | alice29.txt      | 1,727.9 us |  8.53 us |  7.98 us | 1,723.9 us | 1,719.6 us | 1,747.2 us |  1.00 | Same            |    0.00 |   8.26 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Decompress | e9e33e1 (Old) | Fastest | alice29.txt      |   579.8 us |  1.49 us |  1.32 us |   579.5 us |   578.0 us |   582.5 us |  1.00 | Base            |    0.00 |   8.27 KB |        1.00 |
| Decompress | e5179e7 (New) | Fastest | alice29.txt      |   579.2 us |  2.20 us |  1.95 us |   579.3 us |   576.1 us |   582.6 us |  1.00 | Same            |    0.00 |   8.27 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Compress   | e9e33e1 (Old) | Fastest | sum              |   317.8 us |  8.60 us |  9.91 us |   318.0 us |   306.5 us |   335.7 us |  1.00 | Base            |    0.04 |   8.26 KB |        1.00 |
| Compress   | e5179e7 (New) | Fastest | sum              |   321.0 us |  7.35 us |  8.46 us |   323.3 us |   308.2 us |   336.6 us |  1.01 | Same            |    0.04 |   8.26 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Decompress | e9e33e1 (Old) | Fastest | sum              |   150.0 us |  3.25 us |  3.75 us |   151.3 us |   145.4 us |   157.7 us |  1.00 | Base            |    0.03 |   8.27 KB |        1.00 |
| Decompress | e5179e7 (New) | Fastest | sum              |   150.6 us |  2.23 us |  1.86 us |   151.3 us |   146.4 us |   153.2 us |  1.00 | Same            |    0.03 |   8.27 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Compress   | e9e33e1 (Old) | Fastest | TestDocument.pdf | 2,521.2 us |  9.59 us |  8.50 us | 2,517.9 us | 2,513.2 us | 2,539.2 us |  1.00 | Base            |    0.00 |   8.26 KB |        1.00 |
| Compress   | e5179e7 (New) | Fastest | TestDocument.pdf | 2,518.9 us |  8.55 us |  7.58 us | 2,517.9 us | 2,511.0 us | 2,536.9 us |  1.00 | Same            |    0.00 |   8.26 KB |        1.00 |
|            |            |                                                                                                            |         |                  |            |          |          |            |            |            |       |                 |         |           |             |
| Decompress | e9e33e1 (Old) | Fastest | TestDocument.pdf |   393.9 us |  1.01 us |  0.89 us |   393.7 us |   392.9 us |   396.0 us |  1.00 | Base            |    0.00 |   8.27 KB |        1.00 |
| Decompress | e5179e7 (New) | Fastest | TestDocument.pdf |   393.8 us |  0.70 us |  0.59 us |   393.7 us |   393.0 us |   395.0 us |  1.00 | Same            |    0.00 |   8.27 KB |        1.00 |

**Comment by carlossanlop** [comment]
@ericstj @GrabYourPitchforks @artl93 here are the results in Windows arm64, which actually consumes zlib (not zlib-intel). There wasn't any significant perf difference.

Note that new is top, base is bottom:

| Method     | Commit        | level   | file             | Mean       | Error     | StdDev    | Median     | Min        | Max        | Ratio | MannWhitney(5%) | RatioSD | Gen0   | Allocated | Alloc Ratio |
|----------- |-------------- |-------- |-----------------:|-----------:|----------:|----------:|-----------:|-----------:|-----------:|------ |----------------:|--------:|-------:|-----------|------------:|
| Compress   | e5179e7 (new) | Optimal | alice29.txt      | 9,931.6 us |  22.07 us |  20.64 us | 9,936.1 us | 9,891.2 us | 9,959.9 us |  1.00 | Same            |    0.00 |      - |   8.26 KB |        1.00 |
| Compress   | e9e33e1 (old) | Optimal | alice29.txt      | 9,932.4 us |  14.51 us |  12.12 us | 9,937.3 us | 9,911.7 us | 9,946.8 us |  1.00 | Base            |    0.00 |      - |   8.26 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Decompress | e5179e7 (new) | Optimal | alice29.txt      |   615.3 us |   2.59 us |   2.30 us |   615.1 us |   611.9 us |   620.4 us |  1.00 | Same            |    0.01 | 2.5000 |   8.27 KB |        1.00 |
| Decompress | e9e33e1 (old) | Optimal | alice29.txt      |   614.9 us |   2.51 us |   2.22 us |   614.7 us |   611.0 us |   618.4 us |  1.00 | Base            |    0.00 | 2.5000 |   8.27 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Compress   | e5179e7 (new) | Optimal | sum              | 2,313.6 us |   7.69 us |   6.82 us | 2,312.8 us | 2,302.7 us | 2,325.7 us |  1.00 | Same            |    0.00 |      - |   8.26 KB |        1.00 |
| Compress   | e9e33e1 (old) | Optimal | sum              | 2,317.9 us |   8.45 us |   7.91 us | 2,316.5 us | 2,307.0 us | 2,333.2 us |  1.00 | Base            |    0.00 |      - |   8.26 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Decompress | e5179e7 (new) | Optimal | sum              |   158.4 us |   0.62 us |   0.58 us |   158.5 us |   156.8 us |   159.2 us |  1.01 | Same            |    0.01 | 3.7500 |   8.27 KB |        1.00 |
| Decompress | e9e33e1 (old) | Optimal | sum              |   157.6 us |   1.01 us |   0.95 us |   157.6 us |   155.8 us |   159.3 us |  1.00 | Base            |    0.01 | 3.7500 |   8.27 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Compress   | e5179e7 (new) | Optimal | TestDocument.pdf | 4,182.0 us |   9.11 us |   8.07 us | 4,179.4 us | 4,169.1 us | 4,198.6 us |  1.00 | Same            |    0.00 |      - |   8.26 KB |        1.00 |
| Compress   | e9e33e1 (old) | Optimal | TestDocument.pdf | 4,187.7 us |  16.48 us |  14.61 us | 4,183.1 us | 4,168.5 us | 4,220.2 us |  1.00 | Base            |    0.00 |      - |   8.26 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Decompress | e5179e7 (new) | Optimal | TestDocument.pdf |   476.7 us |   1.28 us |   1.14 us |   476.6 us |   474.7 us |   478.8 us |  1.00 | Same            |    0.00 | 3.9063 |   8.27 KB |        1.00 |
| Decompress | e9e33e1 (old) | Optimal | TestDocument.pdf |   478.2 us |   1.76 us |   1.65 us |   478.3 us |   475.1 us |   480.8 us |  1.00 | Base            |    0.00 | 3.9063 |   8.27 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Compress   | e5179e7 (new) | Fastest | alice29.txt      | 2,953.2 us |  16.97 us |  14.17 us | 2,954.9 us | 2,925.8 us | 2,980.8 us |  0.92 | Same            |    0.07 |      - |   8.26 KB |        1.00 |
| Compress   | e9e33e1 (old) | Fastest | alice29.txt      | 3,238.6 us | 222.71 us | 247.54 us | 3,182.8 us | 2,944.8 us | 3,796.3 us |  1.01 | Base            |    0.10 |      - |   8.26 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Decompress | e5179e7 (new) | Fastest | alice29.txt      |   654.6 us |   1.94 us |   1.62 us |   654.6 us |   650.9 us |   656.5 us |  1.00 | Same            |    0.00 | 2.6042 |   8.27 KB |        1.00 |
| Decompress | e9e33e1 (old) | Fastest | alice29.txt      |   656.7 us |   2.55 us |   2.39 us |   656.6 us |   652.7 us |   661.1 us |  1.00 | Base            |    0.00 | 2.7174 |   8.27 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Compress   | e5179e7 (new) | Fastest | sum              |   629.8 us |   5.34 us |   4.99 us |   630.4 us |   618.4 us |   637.0 us |  0.99 | Same            |    0.01 | 2.5000 |   8.26 KB |        1.00 |
| Compress   | e9e33e1 (old) | Fastest | sum              |   633.6 us |   7.43 us |   6.95 us |   634.9 us |   617.1 us |   641.9 us |  1.00 | Base            |    0.02 | 2.5000 |   8.26 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Decompress | e5179e7 (new) | Fastest | sum              |   170.9 us |   1.13 us |   1.00 us |   170.9 us |   169.0 us |   172.6 us |  1.00 | Same            |    0.01 | 3.3967 |   8.27 KB |        1.00 |
| Decompress | e9e33e1 (old) | Fastest | sum              |   170.2 us |   0.90 us |   0.84 us |   169.9 us |   169.2 us |   171.9 us |  1.00 | Base            |    0.01 | 4.0323 |   8.27 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Compress   | e5179e7 (new) | Fastest | TestDocument.pdf | 4,057.5 us |  21.44 us |  20.05 us | 4,061.9 us | 4,024.1 us | 4,090.2 us |  1.00 | Same            |    0.01 |      - |   8.26 KB |        1.00 |
| Compress   | e9e33e1 (old) | Fastest | TestDocument.pdf | 4,041.7 us |   8.36 us |   7.82 us | 4,040.0 us | 4,027.0 us | 4,052.8 us |  1.00 | Base            |    0.00 |      - |   8.26 KB |        1.00 |
|            |               |         |                  |            |           |           |            |            |            |       |                 |         |        |           |             |
| Decompress | e5179e7 (new) | Fastest | TestDocument.pdf |   481.7 us |   2.11 us |   1.76 us |   481.4 us |   479.3 us |   485.8 us |  1.00 | Same            |    0.00 | 3.9063 |   8.27 KB |        1.00 |
| Decompress | e9e33e1 (old) | Fastest | TestDocument.pdf |   481.5 us |   1.95 us |   1.73 us |   480.9 us |   478.7 us |   485.3 us |  1.00 | Base            |    0.00 | 3.9063 |   8.27 KB |        1.00 |

**Comment by ericstj** [comment]
Looks good.  We can also double-check the results from the perf-lab since this is merged now.  Looks like we have some ZLIB benchmarks that will run on more platforms.  I think that's https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/TestHistoryIndexIndex.html

**Comment by carlossanlop** [comment]
Sure, I can also check those results. First I need to understand if they get executed periodically in an automated way or if I need to execute them myself manually.

Since the servicing branches are currently closed, would anyone object to merging the servicing PRs so they bake in staging until the branches open again?

**Pull request merged**